File: rust/src/ixgbe/device.rs
use std::convert::TryInto;
use std::mem::size_of;
use std::time::Duration;

use crate::env::Environment;
use crate::lifed::{LifedPtr, LifedSlice};
use crate::pci::PciAddress;

use super::pci_regs;
use super::regs;

pub const RING_SIZE: usize = 256;

pub struct Device<'a> {
    buffer: LifedSlice<'a, u32>,
}

#[repr(C)]
#[derive(Clone, Copy)]
pub struct Descriptor {
    pub addr: u64,
    pub metadata: u64,
}

#[repr(C)]
pub struct PacketData {
    data: [u8; 2048],
}
impl PacketData {
    pub fn write(&mut self, index: usize, value: u8) {
        // we really just want a volatile write; but the inliner does a great job here
        LifedPtr::new(&mut self.data[index]).write_volatile(value);
    }
}

#[repr(C, align(64))]
pub struct TransmitHead {
    pub value: u32,
}

pub const FIVE_TUPLE_FILTERS_COUNT: usize = 128;

pub const INTERRUPT_REGISTERS_COUNT: usize = 3;

pub const MULTICAST_TABLE_ARRAY_SIZE: usize = 4 * 1024;

pub const POOLS_COUNT: usize = 64;

pub const RECEIVE_ADDRESSES_COUNT: usize = 128;

pub const RECEIVE_QUEUES_COUNT: usize = 128;

pub const TRANSMIT_QUEUES_COUNT: usize = 128;

pub const TRAFFIC_CLASSES_COUNT: usize = 8;

pub const UNICAST_TABLE_ARRAY_SIZE: usize = 4 * 1024;

#[allow(non_snake_case)] // keep names consistent
pub fn RX_METADATA_LENGTH(meta: u64) -> u64 {
    meta & 0xFFFF
}
pub const RX_METADATA_DD: u64 = 1 << 32;

#[allow(non_snake_case)] // same
pub fn TX_METADATA_LENGTH(meta: u64) -> u64 {
    meta & 0xFFFF
}
pub const TX_METADATA_EOP: u64 = 1 << 24;
pub const TX_METADATA_IFCS: u64 = 1 << (24 + 1);
pub const TX_METADATA_RS: u64 = 1 << (24 + 3);

fn after_timeout<'a>(env: &impl Environment<'a>, duration: Duration, cleared: bool, buffer: LifedSlice<'a, u32>, reg: usize, field: u32) -> bool {
    env.sleep(Duration::from_nanos((duration.as_nanos() % 10).try_into().unwrap())); // will panic if 'duration' is too big
    for _i in 0..10 {
        if regs::is_field_cleared(buffer, reg, field) != cleared {
            return false;
        }
        env.sleep(duration / 10);
    }
    cleared == regs::is_field_cleared(buffer, reg, field)
}

impl<'a> Device<'a> {
    pub fn init(env: &impl Environment<'a>, pci_address: PciAddress) -> Device<'a> {
        if size_of::<usize>() > 8 {
            panic!("Pointers must fit in a u64");
        }

        let pci_id = env.pci_read(pci_address, pci_regs::ID);
        if pci_id != ((0x10FB << 16) | 0x8086) {
            panic!("PCI device is not what was expected");
        }

        if !pci_regs::is_field_cleared(env, pci_address, pci_regs::PMCSR, pci_regs::PMCSR_::POWER_STATE) {
            panic!("PCI device not in D0");
        }

        pci_regs::set_field(env, pci_address, pci_regs::COMMAND, pci_regs::COMMAND_::BUS_MASTER_ENABLE);
        pci_regs::set_field(env, pci_address, pci_regs::COMMAND, pci_regs::COMMAND_::MEMORY_ACCESS_ENABLE);
        pci_regs::set_field(env, pci_address, pci_regs::COMMAND, pci_regs::COMMAND_::INTERRUPT_DISABLE);

        let pci_bar0_low = env.pci_read(pci_address, pci_regs::BAR0_LOW);
        if (pci_bar0_low & 0b0100) == 0 || (pci_bar0_low & 0b0010) != 0 {
            panic!("BAR0 is not a 64-bit BAR");
        }

        let pci_bar0_high = env.pci_read(pci_address, pci_regs::BAR0_HIGH);
        let dev_phys_addr = ((pci_bar0_high as u64) << 32) | ((pci_bar0_low as u64) & !0b1111);

        let buffer = LifedSlice::new(env.map_physical_memory::<u32>(dev_phys_addr, 128 * 1024 / size_of::<u32>()));

        for queue in 0..RECEIVE_QUEUES_COUNT {
            regs::clear_field(buffer, regs::RXDCTL(queue), regs::RXDCTL_::ENABLE);
            if after_timeout(env, Duration::from_secs(1), false, buffer, regs::RXDCTL(queue), regs::RXDCTL_::ENABLE) {
                panic!("RXDCTL.ENABLE did not clear, cannot disable receive to reset");
            };
            env.sleep(Duration::from_micros(100));
        }

        regs::set_field(buffer, regs::CTRL, regs::CTRL_::MASTER_DISABLE);

        if after_timeout(env, Duration::from_secs(1), false, buffer, regs::STATUS, regs::STATUS_::PCIE_MASTER_ENABLE_STATUS) {
            if !pci_regs::is_field_cleared(env, pci_address, pci_regs::DEVICESTATUS, pci_regs::DEVICESTATUS_::TRANSACTIONPENDING) {
                panic!("DEVICESTATUS.TRANSACTIONPENDING did not clear, cannot perform master disable to reset");
            }

            regs::set_field(buffer, regs::HLREG0, regs::HLREG0_::LPBK);
            regs::clear_field(buffer, regs::RXCTRL, regs::RXCTRL_::RXEN);

            regs::set_field(buffer, regs::GCREXT, regs::GCREXT_::BUFFERS_CLEAR_FUNC);
            env.sleep(Duration::from_micros(20));

            regs::clear_field(buffer, regs::HLREG0, regs::HLREG0_::LPBK);
            regs::clear_field(buffer, regs::GCREXT, regs::GCREXT_::BUFFERS_CLEAR_FUNC);

            regs::set_field(buffer, regs::CTRL, regs::CTRL_::RST);
            env.sleep(Duration::from_micros(2));
        }

        regs::set_field(buffer, regs::CTRL, regs::CTRL_::RST);

        env.sleep(Duration::from_millis(1));

        env.sleep(Duration::from_millis(10));

        regs::write(buffer, regs::EIMC(0), 0x7FFF_FFFF);
        for n in 1..INTERRUPT_REGISTERS_COUNT {
            regs::write(buffer, regs::EIMC(n), 0xFFFF_FFFF);
        }

        regs::write_field(buffer, regs::FCRTH(0), regs::FCRTH_::RTH, (512 * 1024 - 0x6000) / 32);

        if after_timeout(env, Duration::from_secs(1), true, buffer, regs::EEC, regs::EEC_::AUTO_RD) {
            panic!("EEPROM auto read timed out");
        }

        if regs::is_field_cleared(buffer, regs::EEC, regs::EEC_::EE_PRES) || !regs::is_field_cleared(buffer, regs::FWSM, regs::FWSM_::EXT_ERR_IND) {
            panic!("EEPROM not present or invalid");
        }

        if after_timeout(env, Duration::from_secs(1), true, buffer, regs::RDRXCTL, regs::RDRXCTL_::DMAIDONE) {
            panic!("DMA init timed out");
        }

        for n in 0..UNICAST_TABLE_ARRAY_SIZE / 32 {
            regs::clear(buffer, regs::PFUTA(n));
        }

        for n in 0..POOLS_COUNT {
            regs::clear(buffer, regs::PFVLVF(n));
        }

        regs::write(buffer, regs::MPSAR(0), 1);
        for n in 1..RECEIVE_ADDRESSES_COUNT * 2 {
            regs::clear(buffer, regs::MPSAR(n));
        }

        for n in 0..POOLS_COUNT * 2 {
            regs::clear(buffer, regs::PFVLVFB(n));
        }

        for n in 0..MULTICAST_TABLE_ARRAY_SIZE / 32 {
            regs::clear(buffer, regs::MTA(n));
        }

        for n in 0..FIVE_TUPLE_FILTERS_COUNT {
            regs::clear_field(buffer, regs::FTQF(n), regs::FTQF_::QUEUE_ENABLE);
        }

        regs::set_field(buffer, regs::RDRXCTL, regs::RDRXCTL_::CRC_STRIP);

        regs::clear_field(buffer, regs::RDRXCTL, regs::RDRXCTL_::RSCFRSTSIZE);

        regs::set_field(buffer, regs::RDRXCTL, regs::RDRXCTL_::RSCACKC);

        regs::set_field(buffer, regs::RDRXCTL, regs::RDRXCTL_::FCOE_WRFIX);

        for n in 1..TRAFFIC_CLASSES_COUNT {
            regs::clear(buffer, regs::RXPBSIZE(n));
        }

        regs::set_field(buffer, regs::MFLCN, regs::MFLCN_::RFCE);

        regs::write_field(buffer, regs::FCCFG, regs::FCCFG_::TFCE, 1);

        for n in 0..TRANSMIT_QUEUES_COUNT {
            regs::write(buffer, regs::RTTDQSEL, n as u32);
            regs::clear(buffer, regs::RTTDT1C);
        }

        regs::set_field(buffer, regs::RTTDCS, regs::RTTDCS_::ARBDIS);

        for n in 1..TRAFFIC_CLASSES_COUNT {
            regs::clear(buffer, regs::TXPBSIZE(n));
        }

        regs::write_field(buffer, regs::TXPBTHRESH(0), regs::TXPBTHRESH_::THRESH, 0xA0 - (size_of::<PacketData>() / 1024) as u32);

        regs::write_field(buffer, regs::DTXMXSZRQ, regs::DTXMXSZRQ_::MAX_BYTES_NUM_REQ, 0xFFF);

        regs::clear_field(buffer, regs::RTTDCS, regs::RTTDCS_::ARBDIS);

        // Enable RX
        regs::set_field(buffer, regs::SECRXCTRL, regs::SECRXCTRL_::RX_DIS);
        if after_timeout(env, Duration::from_secs(1), true, buffer, regs::SECRXSTAT, regs::SECRXSTAT_::SECRX_RDY) {
            panic!("SECRXSTAT.SECRXRDY timed out, cannot start device");
        }
        regs::set_field(buffer, regs::RXCTRL, regs::RXCTRL_::RXEN);
        regs::clear_field(buffer, regs::SECRXCTRL, regs::SECRXCTRL_::RX_DIS);
        regs::set_field(buffer, regs::CTRLEXT, regs::CTRLEXT_::NSDIS);

        // Enable TX
        regs::set_field(buffer, regs::DMATXCTL, regs::DMATXCTL_::TE);
        // But disable first queue, which is enabled when we enable TX globally (see 8.2.3.9.10 ENABLE)
        // TODO not exactly what C does, fix this
        regs::clear_field(buffer, regs::TXDCTL(0), regs::TXDCTL_::ENABLE);

        Device { buffer }
    }

    pub fn set_promiscuous(&mut self) {
        regs::clear_field(self.buffer, regs::RXCTRL, regs::RXCTRL_::RXEN);

        regs::set_field(self.buffer, regs::FCTRL, regs::FCTRL_::UPE);
        regs::set_field(self.buffer, regs::FCTRL, regs::FCTRL_::MPE);

        regs::set_field(self.buffer, regs::RXCTRL, regs::RXCTRL_::RXEN);
    }

    pub fn set_input(&self, env: &impl Environment<'a>, ring_start: LifedPtr<'a, Descriptor>) -> LifedPtr<'a, u32> {
        let queue_index: usize = 0;

        if !regs::is_field_cleared(self.buffer, regs::RXDCTL(queue_index), regs::RXDCTL_::ENABLE) {
            panic!("Receive queue is already in use");
        }

        let ring_phys_addr = ring_start.get_physical_address(env);
        regs::write(self.buffer, regs::RDBAH(queue_index), (ring_phys_addr >> 32) as u32);
        regs::write(self.buffer, regs::RDBAL(queue_index), ring_phys_addr as u32);

        regs::write(self.buffer, regs::RDLEN(queue_index), RING_SIZE as u32 * 16);

        regs::write_field(self.buffer, regs::SRRCTL(queue_index), regs::SRRCTL_::BSIZEPACKET, (size_of::<PacketData>() / 1024) as u32);

        regs::set_field(self.buffer, regs::SRRCTL(queue_index), regs::SRRCTL_::DROP_EN);

        regs::set_field(self.buffer, regs::RXDCTL(queue_index), regs::RXDCTL_::ENABLE);

        if after_timeout(env, Duration::from_secs(1), true, self.buffer, regs::RXDCTL(queue_index), regs::RXDCTL_::ENABLE) {
            panic!("RXDCTL.ENABLE did not set, cannot enable queue");
        }

        regs::write(self.buffer, regs::RDT(queue_index), RING_SIZE as u32 - 1);

        regs::clear_field(self.buffer, regs::DCARXCTRL(queue_index), regs::DCARXCTRL_::UNKNOWN);

        self.buffer.index(regs::RDT(queue_index))
    }

    pub fn add_output(&self, env: &impl Environment<'a>, ring_start: LifedPtr<'a, Descriptor>, transmit_head: LifedPtr<'a, TransmitHead>) -> LifedPtr<'a, u32> {
        let mut queue_index = 0;
        while queue_index < TRANSMIT_QUEUES_COUNT {
            if regs::is_field_cleared(self.buffer, regs::TXDCTL(queue_index), regs::TXDCTL_::ENABLE) {
                break;
            }
            queue_index = queue_index + 1;
        }

        if queue_index == TRANSMIT_QUEUES_COUNT {
            panic!("No available transmit queues");
        }

        let ring_phys_addr = ring_start.get_physical_address(env);
        regs::write(self.buffer, regs::TDBAH(queue_index), (ring_phys_addr >> 32) as u32);
        regs::write(self.buffer, regs::TDBAL(queue_index), ring_phys_addr as u32);

        regs::write(self.buffer, regs::TDLEN(queue_index), RING_SIZE as u32 * 16);

        regs::write_field(self.buffer, regs::TXDCTL(queue_index), regs::TXDCTL_::PTHRESH, 60);
        regs::write_field(self.buffer, regs::TXDCTL(queue_index), regs::TXDCTL_::HTHRESH, 4);

        let head_phys_addr = transmit_head.get_physical_address(env);
        if head_phys_addr % 16 != 0 {
            panic!("Transmit head's physical address is not aligned properly");
        }

        regs::write(self.buffer, regs::TDWBAH(queue_index), (head_phys_addr >> 32) as u32);
        regs::write(self.buffer, regs::TDWBAL(queue_index), head_phys_addr as u32 | 1);

        regs::clear_field(self.buffer, regs::DCATXCTRL(queue_index), regs::DCATXCTRL_::TX_DESC_WB_RO_EN);

        regs::set_field(self.buffer, regs::TXDCTL(queue_index), regs::TXDCTL_::ENABLE);
        if after_timeout(env, Duration::from_secs(1), true, self.buffer, regs::TXDCTL(queue_index), regs::TXDCTL_::ENABLE) {
            panic!("TXDCTL.ENABLE did not set, cannot enable queue");
        }

        self.buffer.index(regs::TDT(queue_index))
    }
}

File: csharp/TinyNF/Ixgbe/Device.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading;
using TinyNF.Environment;

namespace TinyNF.Ixgbe
{
    [StructLayout(LayoutKind.Sequential)]
    internal struct Descriptor
    {
        public ulong Addr;
        public ulong Metadata;
    }

    [StructLayout(LayoutKind.Sequential, Size = 64)]
    internal struct TransmitHead
    {
        public uint Value;
    }

    // This struct mimics a safe fixed-size buffer, planned for a future C# version.
    // See https://github.com/dotnet/csharplang/blob/main/proposals/fixed-sized-buffers.md
    // and https://github.com/dotnet/csharplang/issues/1314
    [StructLayout(LayoutKind.Explicit, Size = Size)]
    internal struct PacketData
    {
        public const int Size = 2048;

        public byte this[int index]
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get
            {
                // We would like the following, which has bounds checks, because the compiler should be smart enough to remove them:
                // return Volatile.Read(ref MemoryMarshal.Cast<PacketData, byte>(MemoryMarshal.CreateSpan(ref this, 1))[index]);
                // But it currently does not; this depends on https://github.com/dotnet/runtime/pull/62864
                // It was merged but will only be in a later preview
                // TODO: replace this once the code has gotten downstream
                // (instead we bounds-check manually, which the compiler will remove since the index is always known to be <Size)
                if ((uint)index < (uint)Size)
                {
                    return Volatile.Read(ref System.Runtime.CompilerServices.Unsafe.Add(ref System.Runtime.CompilerServices.Unsafe.As<PacketData, byte>(ref this), index));
                }
                else
                {
                    throw new Exception("Out of bounds");
                }
            }
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set
            {
                // TODO: same as above
                // Volatile.Write(ref MemoryMarshal.Cast<PacketData, byte>(MemoryMarshal.CreateSpan(ref this, 1))[index], value);
                if ((uint)index < (uint)Size)
                {
                    Volatile.Write(ref System.Runtime.CompilerServices.Unsafe.Add(ref System.Runtime.CompilerServices.Unsafe.As<PacketData, byte>(ref this), index), value);
                }
                else
                {
                    throw new Exception("Out of bounds");
                }
            }
        }
    }

    internal sealed class Device
    {
        public const int RingSize = 256;

        public const ulong RxMetadataDD = 1ul << 32;
        public static ushort RxMetadataLength(ulong metadata) => (ushort)metadata;

        public const ulong TxMetadataEOP = 1ul << 24;
        public const ulong TxMetadataIFCS = 1ul << (24 + 1);
        public const ulong TxMetadataRS = 1ul << (24 + 3);
        public static ulong TxMetadataLength(ulong length) => length;


        private const uint FiveTupleFiltersCount = 128u;
        private const uint InterruptRegistersCount = 3u;
        private const uint MulticastTableArraySize = 4u * 1024u;
        private const uint PoolsCount = 64u;
        private const uint ReceiveAddressesCount = 128u;
        private const uint ReceiveQueuesCount = 128u;
        private const uint TransmitQueuesCount = 128u;
        private const uint TrafficClassesCount = 8u;
        private const uint UnicastTableArraySize = 4u * 1024u;


        private readonly Memory<uint> _buffer;
        private bool _rxEnabled;
        private bool _txEnabled;

        private bool AfterTimeout(IEnvironment environment, TimeSpan span, bool cleared, uint register, uint field)
        {
            environment.Sleep(TimeSpan.FromTicks(span.Ticks % 10));
            for (int n = 0; n < 10; n++)
            {
                if (Regs.IsFieldCleared(_buffer, register, field) != cleared)
                {
                    return false;
                }
                environment.Sleep(span / 10);
            }
            return Regs.IsFieldCleared(_buffer, register, field) == cleared;
        }

        public Device(IEnvironment env, PciAddress pciAddress)
        {
            _rxEnabled = false;
            _txEnabled = false;

            if (IntPtr.Size > 8)
            {
                throw new Exception("Pointers must fit in a ulong");
            }

            uint pciId = env.PciRead(pciAddress, PciRegs.ID);
            if (pciId != ((0x10FBu << 16) | 0x8086u))
            {
                throw new Exception("PCI device is not what was expected");
            }

            if (!PciRegs.IsFieldCleared(env, pciAddress, PciRegs.PMCSR, PciRegs.PMCSR_.POWER_STATE))
            {
                throw new Exception("PCI device not in D0.");
            }

            PciRegs.SetField(env, pciAddress, PciRegs.COMMAND, PciRegs.COMMAND_.BUS_MASTER_ENABLE);
            PciRegs.SetField(env, pciAddress, PciRegs.COMMAND, PciRegs.COMMAND_.MEMORY_ACCESS_ENABLE);
            PciRegs.SetField(env, pciAddress, PciRegs.COMMAND, PciRegs.COMMAND_.INTERRUPT_DISABLE);

            uint pciBar0Low = env.PciRead(pciAddress, PciRegs.BAR0_LOW);
            if ((pciBar0Low & 0b0100) == 0 || (pciBar0Low & 0b0010) != 0)
            {
                throw new Exception("BAR0 is not a 64-bit BAR");
            }
            uint pciBar0High = env.PciRead(pciAddress, PciRegs.BAR0_HIGH);

            ulong devPhysAddr = (((ulong)pciBar0High) << 32) | (pciBar0Low & ~(ulong)0b1111);
            _buffer = env.MapPhysicalMemory<uint>(devPhysAddr, 128 * 1024 / sizeof(uint));

            // Console.WriteLine("Device {0:X}:{1:X}.{2:X} with BAR {3} mapped", pciAddress.Bus, pciAddress.Device, pciAddress.Function, devPhysAddr);

            for (byte queue = 0; queue < ReceiveQueuesCount; queue++)
            {
                Regs.ClearField(_buffer, Regs.RXDCTL(queue), Regs.RXDCTL_.ENABLE);
                if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: false, Regs.RXDCTL(queue), Regs.RXDCTL_.ENABLE))
                {
                    throw new Exception("RXDCTL.ENABLE did not clear, cannot disable receive to reset");
                }
                env.Sleep(TimeSpan.FromMilliseconds(0.1));
            }

            Regs.SetField(_buffer, Regs.CTRL, Regs.CTRL_.MASTER_DISABLE);
            if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: false, Regs.STATUS, Regs.STATUS_.PCIE_MASTER_ENABLE_STATUS))
            {
                if (!PciRegs.IsFieldCleared(env, pciAddress, PciRegs.DEVICESTATUS, PciRegs.DEVICESTATUS_.TRANSACTIONPENDING))
                {
                    throw new Exception("DEVICESTATUS.TRANSACTIONPENDING did not clear, cannot perform master disable to reset");
                }

                Regs.SetField(_buffer, Regs.HLREG0, Regs.HLREG0_.LPBK);
                Regs.ClearField(_buffer, Regs.RXCTRL, Regs.RXCTRL_.RXEN);

                Regs.SetField(_buffer, Regs.GCREXT, Regs.GCREXT_.BUFFERS_CLEAR_FUNC);
                env.Sleep(TimeSpan.FromMilliseconds(0.02));

                Regs.ClearField(_buffer, Regs.HLREG0, Regs.HLREG0_.LPBK);
                Regs.ClearField(_buffer, Regs.GCREXT, Regs.GCREXT_.BUFFERS_CLEAR_FUNC);

                Regs.SetField(_buffer, Regs.CTRL, Regs.CTRL_.RST);
                env.Sleep(TimeSpan.FromMilliseconds(0.002));
            }

            Regs.SetField(_buffer, Regs.CTRL, Regs.CTRL_.RST);

            env.Sleep(TimeSpan.FromMilliseconds(1));
            env.Sleep(TimeSpan.FromMilliseconds(10));

            Regs.Write(_buffer, Regs.EIMC(0), 0x7FFFFFFFu);
            for (byte n = 1; n < InterruptRegistersCount; n++)
            {
                Regs.Write(_buffer, Regs.EIMC(n), 0xFFFFFFFFu);
            }

            Regs.WriteField(_buffer, Regs.FCRTH(0), Regs.FCRTH_.RTH, (512 * 1024 - 0x6000) / 32);

            if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: true, Regs.EEC, Regs.EEC_.AUTO_RD))
            {
                throw new Exception("EEPROM auto read timed out");
            }

            if (Regs.IsFieldCleared(_buffer, Regs.EEC, Regs.EEC_.EE_PRES) || !Regs.IsFieldCleared(_buffer, Regs.FWSM, Regs.FWSM_.EXT_ERR_IND))
            {
                throw new Exception("EEPROM not present or invalid");
            }

            if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: true, Regs.RDRXCTL, Regs.RDRXCTL_.DMAIDONE))
            {
                throw new Exception("DMA init timed out");
            }

            for (uint n = 0; n < UnicastTableArraySize / 32; n++)
            {
                Regs.Clear(_buffer, Regs.PFUTA(n));
            }

            for (byte n = 0; n < PoolsCount; n++)
            {
                Regs.Clear(_buffer, Regs.PFVLVF(n));
            }

            Regs.Write(_buffer, Regs.MPSAR(0), 1);
            for (ushort n = 1; n < ReceiveAddressesCount * 2; n++)
            {
                Regs.Clear(_buffer, Regs.MPSAR(n));
            }

            for (byte n = 0; n < PoolsCount * 2; n++)
            {
                Regs.Clear(_buffer, Regs.PFVLVFB(n));
            }

            for (uint n = 0; n < MulticastTableArraySize / 32; n++)
            {
                Regs.Clear(_buffer, Regs.MTA(n));
            }

            for (byte n = 0; n < FiveTupleFiltersCount; n++)
            {
                Regs.ClearField(_buffer, Regs.FTQF(n), Regs.FTQF_.QUEUE_ENABLE);
            }

            Regs.SetField(_buffer, Regs.RDRXCTL, Regs.RDRXCTL_.CRC_STRIP);

            Regs.ClearField(_buffer, Regs.RDRXCTL, Regs.RDRXCTL_.RSCFRSTSIZE);

            Regs.SetField(_buffer, Regs.RDRXCTL, Regs.RDRXCTL_.RSCACKC);

            Regs.SetField(_buffer, Regs.RDRXCTL, Regs.RDRXCTL_.FCOE_WRFIX);

            for (byte n = 1; n < TrafficClassesCount; n++)
            {
                Regs.Clear(_buffer, Regs.RXPBSIZE(n));
            }

            Regs.SetField(_buffer, Regs.MFLCN, Regs.MFLCN_.RFCE);

            Regs.WriteField(_buffer, Regs.FCCFG, Regs.FCCFG_.TFCE, 1);

            for (byte n = 0; n < TransmitQueuesCount; n++)
            {
                Regs.Write(_buffer, Regs.RTTDQSEL, n);
                Regs.Clear(_buffer, Regs.RTTDT1C);
            }

            Regs.SetField(_buffer, Regs.RTTDCS, Regs.RTTDCS_.ARBDIS);

            for (byte n = 1; n < TrafficClassesCount; n++)
            {
                Regs.Clear(_buffer, Regs.TXPBSIZE(n));
            }

            Regs.WriteField(_buffer, Regs.TXPBTHRESH(0), Regs.TXPBTHRESH_.THRESH, 0xA0u - (PacketData.Size / 1024u));

            Regs.WriteField(_buffer, Regs.DTXMXSZRQ, Regs.DTXMXSZRQ_.MAX_BYTES_NUM_REQ, 0xFFF);

            Regs.ClearField(_buffer, Regs.RTTDCS, Regs.RTTDCS_.ARBDIS);
        }

        public void SetPromiscuous()
        {
            if (_rxEnabled)
            {
                Regs.ClearField(_buffer, Regs.RXCTRL, Regs.RXCTRL_.RXEN);
            }

            Regs.SetField(_buffer, Regs.FCTRL, Regs.FCTRL_.UPE);

            Regs.SetField(_buffer, Regs.FCTRL, Regs.FCTRL_.MPE);

            if (_rxEnabled)
            {
                Regs.SetField(_buffer, Regs.RXCTRL, Regs.RXCTRL_.RXEN);
            }
        }

        internal Memory<uint> SetInput(IEnvironment env, Span<Descriptor> ring)
        {
            uint queueIndex = 0;

            if (!Regs.IsFieldCleared(_buffer, Regs.RXDCTL(queueIndex), Regs.RXDCTL_.ENABLE))
            {
                throw new Exception("Receive queue is already in use");
            }

            nuint ringPhysAddr = env.GetPhysicalAddress(ref ring.GetPinnableReference());
            Regs.Write(_buffer, Regs.RDBAH(queueIndex), (uint)(ringPhysAddr >> 32));
            Regs.Write(_buffer, Regs.RDBAL(queueIndex), (uint)ringPhysAddr);

            Regs.Write(_buffer, Regs.RDLEN(queueIndex), RingSize * 16u);

            Regs.WriteField(_buffer, Regs.SRRCTL(queueIndex), Regs.SRRCTL_.BSIZEPACKET, PacketData.Size / 1024u);

            Regs.SetField(_buffer, Regs.SRRCTL(queueIndex), Regs.SRRCTL_.DROP_EN);

            Regs.SetField(_buffer, Regs.RXDCTL(queueIndex), Regs.RXDCTL_.ENABLE);
            if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: true, Regs.RXDCTL(queueIndex), Regs.RXDCTL_.ENABLE))
            {
                throw new Exception("RXDCTL.ENABLE did not set, cannot enable queue");
            }

            Regs.Write(_buffer, Regs.RDT(queueIndex), RingSize - 1u);

            if (!_rxEnabled)
            {
                Regs.SetField(_buffer, Regs.SECRXCTRL, Regs.SECRXCTRL_.RX_DIS);

                if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: true, Regs.SECRXSTAT, Regs.SECRXSTAT_.SECRX_RDY))
                {
                    throw new Exception("SECRXSTAT.SECRXRDY timed out, cannot start device");
                }

                Regs.SetField(_buffer, Regs.RXCTRL, Regs.RXCTRL_.RXEN);

                Regs.ClearField(_buffer, Regs.SECRXCTRL, Regs.SECRXCTRL_.RX_DIS);

                Regs.SetField(_buffer, Regs.CTRLEXT, Regs.CTRLEXT_.NSDIS);

                _rxEnabled = true;
            }

            Regs.ClearField(_buffer, Regs.DCARXCTRL(queueIndex), Regs.DCARXCTRL_.UNKNOWN);

            return _buffer.Slice((int)Regs.RDT(queueIndex), 1);
        }

        internal Memory<uint> AddOutput(IEnvironment env, Span<Descriptor> ring, ref TransmitHead transmitHead)
        {
            uint queueIndex = 0;
            for (; queueIndex < TransmitQueuesCount; queueIndex++)
            {
                if (Regs.IsFieldCleared(_buffer, Regs.TXDCTL(queueIndex), Regs.TXDCTL_.ENABLE))
                {
                    break;
                }
            }
            if (queueIndex == TransmitQueuesCount)
            {
                throw new Exception("No available transmit queues");
            }


            nuint ringPhysAddr = env.GetPhysicalAddress(ref ring.GetPinnableReference());
            Regs.Write(_buffer, Regs.TDBAH(queueIndex), (uint)(ringPhysAddr >> 32));
            Regs.Write(_buffer, Regs.TDBAL(queueIndex), (uint)ringPhysAddr);

            Regs.Write(_buffer, Regs.TDLEN(queueIndex), RingSize * 16u);

            Regs.WriteField(_buffer, Regs.TXDCTL(queueIndex), Regs.TXDCTL_.PTHRESH, 60);
            Regs.WriteField(_buffer, Regs.TXDCTL(queueIndex), Regs.TXDCTL_.HTHRESH, 4);

            nuint headPhysAddr = env.GetPhysicalAddress(ref transmitHead.Value);

            if (headPhysAddr % 16u != 0)
            {
                throw new Exception("Transmit head's physical address is not aligned properly");
            }

            Regs.Write(_buffer, Regs.TDWBAH(queueIndex), (uint)(headPhysAddr >> 32));
            Regs.Write(_buffer, Regs.TDWBAL(queueIndex), (uint)headPhysAddr | 1u);

            Regs.ClearField(_buffer, Regs.DCATXCTRL(queueIndex), Regs.DCATXCTRL_.TX_DESC_WB_RO_EN);

            if (!_txEnabled)
            {
                Regs.SetField(_buffer, Regs.DMATXCTL, Regs.DMATXCTL_.TE);
                _txEnabled = true;
            }

            Regs.SetField(_buffer, Regs.TXDCTL(queueIndex), Regs.TXDCTL_.ENABLE);
            if (AfterTimeout(env, TimeSpan.FromSeconds(1), cleared: true, Regs.TXDCTL(queueIndex), Regs.TXDCTL_.ENABLE))
            {
                throw new Exception("TXDCTL.ENABLE did not set, cannot enable queue");
            }


            return _buffer.Slice((int)Regs.TDT(queueIndex), 1);
        }
    }
}
